geopandas/geodataframe.py

```diff
@@ -279,6 +279,8 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
 
         to_remove = None
         geo_column_name = self._geometry_column_name
+        if isinstance(self.columns, pd.MultiIndex) and isinstance(geo_column_name, str):
+            geo_column_name = (geo_column_name,) + ("",) * (self.columns.nlevels - 1)
```

Comment:
I tried to have another dig and come up with a better fix, but following what's going on inside pandas gets complicated pretty quickly. This is a variant on what was there before, borrowed from `pd.MultiIndex._validiate_fill_value`, which is part of the call stack from `GeoDataFrame.__setitem__` with multiindex columns and a partial key (e.g. columns `[["location", "location"], ["x", "y"]]` and key `"geometry"` in this context)


With this, `frame[geo_column_name] = level` below actually works properly and `frame[geo_column_name]` is a GeoSeries of geometry dtype instead of object dtype.

Reply:
This solution looks fine to my eyes.